src/common/viz/FigureExtractor.js

```diff
@@ -17,210 +17,334 @@ define(['./Utils'], function (Utils) {
+                points: points,
+                width: nodeInfo.attributes.width,
+                color: nodeInfo.attributes.color
+            };
+
+            return desc;
+        }
+
+        getExecutionId (/* node */) {
+            throw new Error('getExecutionId is not implemented');
+        }
+
+        getGraphNode (/* node */) {
+            throw new Error('getGraphNode is not implemented');
+        }
+
+        _getContainmentParentNodeAt (/* node, metaType */) {
+            throw new Error('_getContainmentParentNodeAt is not implemented');
+        }
+
+        getMetaType (/* node */) {
+            throw new Error('getMetaType is not implmented');
+        }
+
+        GMENodeToMetadataJSON (/* node, shallow=false */) {
+            throw new Error('GMENodeTOMetadataJSON is not implemented');
+        }
+    }
+
+    class ClientFigureExtractor extends AbstractFigureExtractor {
+        constructor(client) {
+            super();
+            this._client = client;
+        }
+
+        getExecutionId (node) {
+            const executionNode = this._getContainmentParentNodeAt(node, 'Execution');
+            if (executionNode){
+                return executionNode.getId();
+            }
+        }
+
+        getMetadataChildrenIds (node) {
+            const allMetaNodes = this._client.getAllMetaNodes();
+            const metadataBaseNode = allMetaNodes
+                .find(node => node.getAttribute('name') === BASE_METADATA_TYPE);
+
+            if(metadataBaseNode) {
+                return node.getChildrenIds().filter(id => {
+                    return this._client.isTypeOf(id, metadataBaseNode.getId());
+                });
+            } else {
+                return [];
+            }
+        }
+
+        getGraphNode (node) {
+            return this._getContainmentParentNodeAt(node, 'Graph');
+        }
+
+        _getContainmentParentNodeAt (node, metaType) {
+            let currentNode = node,
+                parentId = currentNode.getParentId();
+            const isMetaType = node => this.getMetaType(node) === metaType;
+            while (parentId !== null && !isMetaType(currentNode)) {
+                currentNode = this._client.getNode(parentId);
+                parentId = currentNode.getParentId();
+            }
+            return isMetaType(currentNode) ? currentNode : null;
+        }
+
+        getMetaType (node) {
+            const metaTypeId = node.getMetaTypeId();
+            const metaNode = this._client.getNode(metaTypeId);
+            if(metaNode) {
+                return metaNode.getAttribute('name');
+            }
+        }
+
+        GMENodeToMetadataJSON (node, shallow=false) {
+            const json = {
+                id: node.getId(),
+                parent_id: node.getParentId(),
+                execution_id: this.getExecutionId(node),
+                meta_type: this.getMetaType(node),
+                attributes: {},
+                attributes_meta: {}
```

Comment:
Is `attributes_meta` necessary? Also, it might be nice to not include `execution_id`, `parent_id`, or `meta_type` and instead simply have the `base` and `parent` which would be the result of calling `toJSON` (shallow) on the parent and base.

Reply:
Only `attributes` are extracted.